---
title: "Resarch"
permalink: /research/
author_profile: true
---

## MetaSRA: Normalized sample-specific metadata for the Sequence Read Archive

The NCBI’s Sequence Read Archive (SRA) promises great biological insight if one could analyze the data in the aggregate; however, the data remain largely underutilized, in part, due to the poor structure of the metadata associated with each sample. The rules governing submissions to the SRA do not dictate a standardized set of terms that should be used to describe the biological samples from which the sequencing data are derived. As a result, the metadata include many synonyms, spelling variants and references to outside sources of information.

The NCBI’s Sequence Read Archive (SRA) promises great biological insight if one could analyze the data in the aggregate; 
however, the data remain largely underutilized, in part, due to the poor structure of the metadata associated with each sample. We present MetaSRA, a database of normalized SRA human sample-specific metadata following a schema inspired by the 
metadata organization of the ENCODE project.

The MetaSRA is available at [metasra.biostat.wisc.edu](metasra.biostat.wisc.edu) via both a searchable web interface and bulk downloads. Software implementing our computational pipeline is available at [http://github.com/deweylab/metasra-pipeline](http://github.com/deweylab/metasra-pipeline).

![alt text](https://mbernste.github.io/images/MetaSRA_overview.png "")

Bernstein,M.N., Doan,A., and C.N.Dewey (2017). [Normalized human sample-specific metadata for the Sequence Read Archive](https://doi.org/10.1093/bioinformatics/btx334). _Bioinformatics_, 33(18), 2914–2923. 

