---
title: "Research"
permalink: /research/
author_profile: true
---

## Hierarchical cell type classification 

Gene expression-based classification of a biological sample’s cell type is an important step in many transcriptomic analyses, including that of annotating cell types in single-cell RNA-seq datasets. We explore the novel application of hierarchical classification algorithms that take into account the graph structure of the Cell Ontology to this task. 

[logo]: https://mbernste.github.io/images/MetaSRA_overview.png "Logo Title Text 2"
![My helpful screenshot]({{ "  https://mbernste.github.io/images/cell_type_classification.png" |   https://mbernste.github.io/images/cell_type_classification.png }})

**Paper:**

Bernstein,M.N. and C.N. Dewey (2019). [Hierarchical cell type classification using mass, heterogeneous RNA-seq data from human primary cells](https://www.biorxiv.org/content/10.1101/634097v1). _bioRxiv_.

## MetaSRA: Normalized metadata for the Sequence Read Archive

The NCBI’s Sequence Read Archive (SRA) promises great biological insight if one could analyze the data in the aggregate; 
however, the data remain largely underutilized, in part, due to the poor structure of the metadata associated with each sample. We present MetaSRA, a database of normalized SRA human sample-specific metadata.  The MetaSRA is available at [metasra.biostat.wisc.edu](metasra.biostat.wisc.edu).

See [my talk](https://www.youtube.com/watch?v=pVHMq9SdUtc) describing the MetaSRA from ISMB 2018.

[logo]: https://mbernste.github.io/images/MetaSRA_overview.png "Logo Title Text 2"
![My helpful screenshot]({{ " https://mbernste.github.io/images/MetaSRA_web.png" |  https://mbernste.github.io/images/MetaSRA_web.png }})

**Paper:**

Bernstein,M.N., Doan,A., and C.N.Dewey (2017). [MetaSRA: Normalized human sample-specific metadata for the Sequence Read Archive](https://doi.org/10.1093/bioinformatics/btx334). _Bioinformatics_, 33(18), 2914–2923. 



