---
title: "MetaSRA: normalized human sample-specific metadata for the Sequence Read Archive"
collection: publications
permalink: /publication/metasra_2017
excerpt: 'This paper is about the number 1. The number 2 is left for future work.'
date: 2009-10-01
venue: 'Journal 1'
paperurl: 'http://academicpages.github.io/files/paper1.pdf'
citation: 'Your Name, You. (2009). &quot;Paper Title Number 1.&quot; <i>Journal 1</i>. 1(1).'
---
The NCBIâ€™s Sequence Read Archive (SRA) promises great biological insight if one could analyze the data in the aggregate; 
however, the data remain largely underutilized, in part, due to the poor structure of the metadata associated with each sample. 
The rules governing submissions to the SRA do not dictate a standardized set of terms that should be used to describe the biological 
samples from which the sequencing data are derived. As a result, the metadata include many synonyms, spelling variants and 
references to outside sources of information. Furthermore, manual annotation of the data remains intractable due to the large number of samples in the archive. For these reasons, it has been difficult to perform large-scale analyses that study the relationships between biomolecular processes and phenotype across diverse diseases, tissues and cell types present in the SRA.
Results: We present MetaSRA, a database of normalized SRA human sample-specific metadata following a schema inspired by the 
metadata organization of the ENCODE project. This schema involves mapping samples to terms in biomedical ontologies, 
labeling each sample with a sample-type category, and extracting real-valued properties. We automated these tasks via a novel 
computational pipeline.

[Download paper here](https://academic.oup.com/bioinformatics/article/3848915/MetaSRA-normalized-human-sample-specific-metadata)

#Recommended citation: Your Name, You. (2009). "Paper Title Number 1." <i>Journal 1</i>. 1(1).
